### ISSUE TITLE
symbols immutability & inconsistent creations

### ISSUE THREAD

**Issue by sylvain-bougnoux** [sympy/sympy#20539]
Hi, I would like to add my own fields, eg foo, to symbols for various purposes such as sorting expression or impose hierarchy in display... But AFAIU symbols are immutable (constant) after creation.

However I realized that at creation time it was a bit different, & inconsistent between Function & Symbol, eg:
```
>f=Function('f',foo=1)
>f.foo
1 
>x=Symbol('x', foo=1)
>x.foo
Traceback (most recent call last):

  File "<ipython-input-112-51c7b810e88e>", line 1, in <module>
    x.foo

AttributeError: 'Symbol' object has no attribute 'foo'
```
Therefore my questions:
1. Is it allowed to add own fields & how? 
2. Is this inconsistency normal?

**Comment by sylvain-bougnoux** [comment]
Actually I just also realized that `Function` objects are not immutable, field can be added or changed anytime, while `Symbol` not. Maybe it is the answer?

**Comment by oscarbenjamin** [comment]
That's because a Function object is really a Python class. Ideally it would be a Basic instance but it isn't https://github.com/sympy/sympy/issues/4787

**Comment by oscarbenjamin** [comment]
It is intended that Symbol and Function should be treated as immutable regardless of whether or not an error is actually raised.

I'm going to close this because there's nothing to fix apart from #4787
